Add support for LDAP user and group in policy attachments

acolombier opened the pull request:
This PR implements the following changes:

- Add support for LDAP DN as `group_name` in the resource `minio_iam_group_policy_attachment`
- Add support for LDAP DN as `user_name` in the resource `minio_iam_user_policy_attachment`

## Reference

 - Resolves: #445

## Closing issues

- Closes: #445

## Note

When running `tfplugindocs` (v0.14.1), it did update couple of syntax unrelated to this change; it now seems to make the difference between optional arguments and read-only arguments.

Currently, I have only pushed those that are relevant to this PR (doc related to the proposed change), but let me know if you would like me to push the other updates in a new PR.

BuJo commented on minio/resource_minio_iam_group_policy_attachment.go:
```diff
@@ -58,9 +58,10 @@ func minioReadGroupPolicyAttachment(ctx context.Context, d *schema.ResourceData,
 
 	groupInfo, errGroup := minioAdmin.GetGroupDescription(ctx, groupName)
-	if errGroup != nil {
+	if errGroup != nil && !isLDAPGroup {
```
I haven't tried it and don't have a docker-compose setup for a LDAP based minio.  So I'm guessing that when using LDAP that `GetGroupDescription` will return an error?

That seems slightly odd..  If it is that way, maybe a better way would be to simply not execute the call if it is an ldap group?

acolombier replied:
Nope it does work just fine if the group actually exists. Here is an example:

```
$ mc admin group info myinstance "CN=my,OU=group,DC=my,DC=org"
Group: CN=my,OU=group,DC=my,DC=orgStatus: 
Policy: 
Members: 
```

Removing this would mean not being able to perform drift detection which would be very problematic IMHO

BuJo replied:
Thank you for the detailed answer!  Trying to wrap my head around it.

If the call to `GetGroupDescription` should work, then I'm unsure what that `if` actually does.

* if the call to `GetGroupDescription` failed AND it's not a normal user THEN return an error from `minioReadGroupPolicyAttachment`

So it ignores failures getting the group description if the group is from LDAP.  Why can we ignore failures at this point?

acolombier replied:
The reason for doing this is that `GetGroupDescription` (and `GetUserInfo`) return isn't consistent. Here is a truth table (note that I'm not sure if `IsZero` is actually defined on groups, but this to represent whether or not the group is a zero value) 

|                               | groupInfo.IsZero | errGroup != nil | group exists in MinIO |
| Get info a for a static group | 0                | 0               | 1                     |
| Get info a for a static group | 1                | 0               | 0                     |
| Get info a for a LDAP group   | 0                | 0               | 1                     |
| Get info a for a LDAP group   | 0                | 1               | 0                     |

Unfortunately, [AdminClient.GetGroupDescription](https://pkg.go.dev/github.com/minio/madmin-go/v2#AdminClient.GetGroupDescription) doesn't describe whether the expect return consistency, so I cannot tell if this is a bug or simply an expected behavior.

Note that exactly the same applies to `GetUserInfo`

I guess one issue of adding is condition, is that it could obfuscate other genuine issue when describing the LDAP group/user (e.g timeout, auth or API misbehavior). I will try and see if I can make this a little bit more resilient

acolombier replied:
The reason for doing this is that `GetGroupDescription` (and `GetUserInfo`) expect the group or the user to be created before hand (e.g using a [minio_iam_group](https://registry.terraform.io/providers/aminueza/minio/latest/docs/resources/iam_group) or a [minio_iam_user](https://registry.terraform.io/providers/aminueza/minio/latest/docs/resources/iam_user)). However, when using a LDAP group or user, you must not create it, but instead should assign a policy right away.

To summarize:
- `GetGroupDescription` or `GetUserInfo` on a static group/user is expected to return a `200` with a list of policies attached to it:
  - If the list is including the policy at hand, it should do nothing (case `all good`, no error)
  - If the list isn't including the policy at hand, it should create (case `missing policy`, no error)
  - If the user/group doesn't exist, it aborts (case `mising user/group`, error)
- `GetGroupDescription` or `GetUserInfo` on a LDAP group/user is expected to return a `200` with a list of policies attached to it:
  - If the list is including the policy at hand, it should do nothing (case `all good`, no error)
  - If the user/group doesn't exist, it should create (case `mising user/group`, no error)

Hopefully, that makes sense now.

An example is provided in `examples/resources/minio_iam_group_policy_attachment/resource.tf` and `examples/resources/minio_iam_user_policy_attachment/resource.tf`

acolombier replied:
I guess one probrem with this `isLDAP...` condition is that it would ignore any API error (e.g: timeout, auth, API misbehavior). I will try and see if I can make that is bit more resilient

BuJo commented on minio/resource_minio_iam_user_policy_attachment.go:
```diff
@@ -55,9 +55,10 @@ func minioCreateUserPolicyAttachment(ctx context.Context, d *schema.ResourceData
 
 	userInfo, errUser := minioAdmin.GetUserInfo(ctx, userName)
-	if errUser != nil {
+	if errUser != nil && !isLDAPUser {
```
I guess that would be exactly the same as for groups - if it will always fail on LDAP groups, it might be better to not execute `GetUserInfo` if minio can't get information about LDAP users.

acolombier replied:
Like for group, this is completely valid in MinIO, as you can see here:

```
$ mc admin user info ttc "CN=Me,OU=User,DC=my,DC=org"
AccessKey: CN=Me,OU=User,DC=my,DC=org
Status: 
PolicyName: consoleAdmin
MemberOf: 
$ mc admin user info ttc "CN=MissingUser,OU=User,DC=my,DC=org"
mc: <ERROR> Unable to get user info. The specified user does not exist. (Specified user does not exist).
```

BuJo reviewed (changes requested):
I had a look at the changes, the only thing that I found a little odd is that minio can't provide information about LDAP users/groups.

acolombier commented:
> I had a look at the changes, the only thing that I found a little odd is that minio can't provide information about LDAP users/groups.

Minio can actually provide information about LDAP users/groups that it has a configuration for. Removing the section that perform LDAP user/group assert removes the ability for the provider to perform drift detection/reconciliation.

acolombier commented:
@BuJo after further digging, it looks like only `GetUserInfo` needs the special error handling in case it is a LDAP user, so I have remove the group one. Also added specific catch for LDAP user, so it wouldn't obfuscate _any_ errors returned by `madmin.AdminClient.GetUserInfo`

Tested locally with a mix set of static users and group, as well as LDAP user and group too. Tested happy path and reconciliation.

I feel like the user custom logic should definitely be covered with unit tests, but AFAIS, this would require some refactoring to support DI in order to allow `madmin.AdminClient.GetUserInfo` mocking.

IMHO, this refactoring should neither happen in this PR as it is out of scope with the task at hand, and unfortunately I'm unsure I currently have the time to put that effort in too. I have added couple of debug output so if this ever regress in future versions, there should at least be some debugging starting point.

Hopefully this is acceptable.

BuJo reviewed (approved):
Allright, I think it's good to go.  While I'd prefer some tests for feeling more comfortable, I'm quite sure it won't break a current usecase.
Thank you for your work!

zeqk commented:
Hi! Thank you very much for work in this feature, I need this.
When will this change be released? 
https://registry.terraform.io/providers/aminueza/minio/latest
https://github.com/aminueza/terraform-provider-minio/releases/

felladrin commented:
Let me make a new release with this change now.

felladrin commented:
> Hi! Thank you very much for work in this feature, I need this. When will this change be released? 

✅ Published in [release 1.14.0](https://registry.terraform.io/providers/aminueza/minio).
